src/edgepi/adc/edgepi_adc.py

```diff
@@ -979,6 +987,136 @@ def set_config(
+            )
+
+            # write config registers, start the adc's conversions, wait, then read registers
+            return (
+                write_reg_cmd + self.adc_ops.start_adc(ADCNum.ADC_1.value),
+                conversion_delay,
+                self.adc_ops.read_adc(ADCNum.ADC_1.value),
+            )
+
+        mux_pairs = (
+            [(channel, CH.AINCOM) for channel in channel_list] +
+            [(diff_mode.value.mux_p, diff_mode.value.mux_n) for diff_mode in differential_pairs]
+        )
+        # get instructions needed to send for each input
+        command_tup_list = [
+            create_read_command_tup(i == 0, mux_p, mux_n)
+            for i, (mux_p, mux_n) in enumerate(mux_pairs)
+        ]
+
+        # TODO: what happens if the SPI read fails (or something else fails in ADC during)?
+        # Register values will be unknown until we read from it again?
+        data_list = self.spi_apply_adc_commands(command_tup_list)
+
+        # TODO: make sure these updates work correctly
+        # update with final ADC state (for state caching)
+        EdgePiADC.__state[ADCReg.REG_MODE2.value] = mode2_register_value
+        mux_p, mux_n = mux_pairs[-1]
+        EdgePiADC.__state[ADCReg.REG_INPMUX.value] = (mux_p << 4) + mux_n
+
+        voltage_list = []
+        for i, read_data in enumerate(data_list):
+            if (len(read_data) - 1) != ADC_VOLTAGE_READ_LEN:
+                raise VoltageReadError(f"Voltage read failed: incorrect number of bytes ({len(read_data)}) retrieved")
+
+            # status_code = read_data[1]
+            voltage_code = read_data[2 : (2 + ADCNum.ADC_1.value.num_data_bytes)]
+            check_code = read_data[6]
+            check_crc(voltage_code, check_code)
+
+            # assert neither mux is set to float mode
+            mux_p, mux_n = mux_pairs[i]
+            if CH.FLOAT in (mux_p, mux_n):
+                raise ValueError("Cannot retrieve calibration values for channel in float mode")
+
+            calib_key = (
+                ADCProperties.ADC1_MUXP.value.values[mux_p.value << 4].value
+                if mux_n == CH.AINCOM
+                else self.__get_diff_id(mux_p, mux_n)
+            )
+            calibs = self.adc_calib_params[ADCNum.ADC_1][calib_key]
+
+            if calibs is None:
+                _logger.error("Failed to find ADC calibration values")
+                raise CalibKeyMissingError(
+                    "Failed to retrieve calibration values from eeprom dictionary: "
+                    f"dict is missing key = {calib_key}"
+                    f"\neeprom_calibs = \n{self.adc_calib_params[ADCNum.ADC_1]}"
+                )
+
+            # convert from voltage_code to voltage value (float)
+            if i < len(channel_list):
+                voltage_list += [code_to_voltage_single_ended(voltage_code, ADCNum.ADC_1.value, calibs)]
+            else:
+                voltage_list += [code_to_voltage(voltage_code, ADCNum.ADC_1.value, calibs)]
```

Comment:
How about separating this into its own method? Maybe something like `build_voltage_list(data_list, mux_pairs)`

Reply:
I decided to break away the calibration parameter generation into `__get_calibration_params_mux` (and update`__get_calibration_params` which we can't use because it reads from the state object [very slow!]), as the rest of the loop body isn't very re-usable.

https://github.com/EdgePi-Cloud/edgepi-python-sdk/pull/424/commits/60d43905982a6a075675af4810442dc81fedf593